File: problems/Java/93. Restore IP Addresses.java
class Solution {
    public List<String> restoreIpAddresses(String s) {
        List<String> res = new ArrayList<>();
        bt(s, 0, new ArrayList<>(), res);
        return res;
    }

    private void bt(String s, int start, List<String> path, List<String> res) {
        if (path.size() == 4 && start == s.length()) {
            res.add(String.join(".", path));
            return;
        }
        if (path.size() >= 4 || start >= s.length()) {
            return;
        }

        for (int i = start; i < start + 3 && i < s.length(); i++) {
            String sub = s.substring(start, i + 1);
            if (sub.length() > 1 && sub.charAt(0) == '0') {
                continue;
            }
            if (Integer.parseInt(sub) <= 255) {
                path.add(sub);
                bt(s, i + 1, path, res);
                path.remove(path.size() - 1);
            }
        }
    }
}

File: problems/Python/93. Restore IP Addresses.py
class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        res = []
        path = []

        def bt(start):
            if len(path) == 4 and start == len(s):
                res.append(".".join(path))
                return
            if len(path) >= 4:
                return

            for i in range(start, min(start + 3, len(s))):
                sub = s[start:i+1]
                if len(sub) > 1 and sub[0] == '0':
                    continue
                if int(sub) <= 255:
                    path.append(sub)
                    bt(i + 1)
                    path.pop()
        
        bt(0)
        return res